### DISCUSSION TITLE
Cart is `null` after `submitOrder`

### DISCUSSION THREAD

**Discussion by vimtor** [swellstores/community#157] [Help] [answered]
Something very odd is happening. After calling the `swell.cart.submitOrder()`, the order is submitted successfully but the method `swell.cart.get()` returns always `null`. This gets solved if I `logout` and `login` again.

I've tried adding new items to the cart, getting a new one, and even creating one and recovering it with the Backend API, but nothing seems to work. I am using the latest Swell SDK version and building a custom storefront using Next.js.

**Comment by vimtor** [comment]
**Update:** Now I cannot add any items no matter what. Sign in again didn't fix the issue this time.

Here are some screenshots of the request after adding an item:

<img width="838" alt="CleanShot 2023-03-02 at 21 30 54@2x" src="https://user-images.githubusercontent.com/36263538/222545062-1bb7d9ec-62ac-464d-b012-544548daa262.png">
<img width="838" alt="CleanShot 2023-03-02 at 21 31 11@2x" src="https://user-images.githubusercontent.com/36263538/222545068-5e7b37b5-27cc-49c8-bc3a-46afa93173a2.png">
<img width="838" alt="CleanShot 2023-03-02 at 21 31 13@2x" src="https://user-images.githubusercontent.com/36263538/222545074-9704d316-d23d-4751-9baa-8ac1cfd40904.png">

**Comment by sags95** [comment]
@vimtor It's expected behaviour for `swell.cart.get()` to return `null` after submitting a cart, as there are no items in this newly created cart (details [here](https://developers.swell.is/frontend-api/carts#get-a-cart)). 

However it is odd that you're getting a 200 but `null` response after adding items to the new cart. Are you perhaps seeing any CORS errors?

**Reply by vimtor** [reply]
Hi @sags95! I am not seeing any CORS error at the moment.

Let me know how can I give you more information to solve this issue. If you want we can schedule a quick meeting to check the code together. Right now this is the only thing blocking us from launching.


https://user-images.githubusercontent.com/36263538/222970077-79633162-4d42-457d-8e54-9f9472d2f428.mp4

**Reply by vimtor** [reply]
Hi there! @sags95 @ericingram

This continues to fail for us. I recorded a video showing you the whole process in case it may help.
At the end of the video I show a way I am able to fix it using the admin SDK and doing some weird stuff that could help you debug the issue.

https://share.cleanshot.com/dXxdK8qj

**Reply by ericingram** [reply]
Thanks for the details @vimtor! As @sags95 said, the first case is normal (returning null after submission), but the other examples do not look normal. Escalating this internally based on your feedback.

**Reply by sags95** [reply]
@vimtor By any chance are you on latest swell-js version?

**Reply by vimtor** [reply]
@sags95  I am on the latest `^3.20.1` version of the `swell-js` package.

**Reply by vimtor** [reply]
Another thing I've noticed (I don't know if it helps) but when I got to the cart page, in which I am updating the cart with the info I show in the video, I get this 400 response. 


<img width="708" alt="CleanShot 2023-03-15 at 23 24 47@2x" src="https://user-images.githubusercontent.com/36263538/225457624-c10ac429-5f62-4bd3-b1d6-d145f1dc994a.png">

**Reply by sags95** [reply]
@vimtor Would you be able to provide a simple stackblitz (or similar) example of your implementation? We've tried to reproduce but are not seeing any errors after adding to cart when not signed in.

**Reply by vimtor** [reply]
@sags95 I invited you to the GitHub repository of the project. If that does not work for you I can try to replicate something similar this weekend in Stackblitz. You can find the logic for everything related to the cart in the `useCart` hook used primarily in the `checkout/cart.tsx` and `add-to-cart-button.tsx` components. Also a small correction, in my store, only logged in users are able to add items to the cart. The project is using Next.js 13 app directory and can be found here https://missing.parts/ you can create an account here by filling random data and going through the add item + checkout process.

Sorry for not being able to provide you with the simple Stackblitz, I don't expect to be able to work on this until next week so I try to send you all the information I can meanwhile. Let me know if you need something else.

Thank you!

**Reply by vimtor** [reply]
Hi @sags95! Do you still need the Stackblitz? From your 👍 I did not understand if with the repository access is enough.
Maybe if you have a guide on how to implement a custom checkout using the `swell-js` I can figure out if I am doing something wrong.

**Reply by sags95** [reply]
Hey @vimtor, thanks for providing access to the repo, I was able to build it locally just fine.

I've spent some time debugging and I feel like SWR is likely doing some caching on the cart. What I might suggest is trying the implementation without SWR to test. Alternatively you can take a look at some of the hooks [here](https://github.com/vercel/commerce/tree/main/packages/swell/src/cart) for some insight on how SWR was implemented in NextJS Commerce.

**Reply by vimtor** [reply]
Hey @sags95, thanks for taking the time to check out the repository! I don't feel like the issue has to do with SWR since it's the API the one returning a `null` cart after trying to add an item. Also, SWR only caches things in JS memory so if it was the problem the issue will resolve after reloading the website. I think I'm going to create another store to see if it is a configuration issue. Again, if you have some guides you can share with me on how to build a custom checkout that would be amazing because right now I had to debug my way through implementation by seeing what methods were available and what errors appeared in the process.

**Reply by vimtor** [reply]
OKEY OKEY I got it working. The issue was that a currency was not selected. I saw this in the request headers:

<img width="430" alt="CleanShot 2023-03-22 at 14 32 50@2x" src="https://user-images.githubusercontent.com/36263538/226920391-b8205bb9-f122-4911-ad01-32e9a338fdd7.png">

I think a guide for custom implementations with all of these details will be highly appreciated because the platform is amazing but I had to debug my was to a complete checkout.

Thank you @sags95 @ericingram and sorry for all the headaches ❤️

**Reply by sags95** [reply]
@vimtor  glad to hear you managed to figure this one out! Your feedback is greatly appreciated, the team is actively working on custom checkout guides so I'll be sure to share as soon as we get these published.

**Comment by vimtor** [comment, marked as answer]
I was missing the following line of code:

```js
await swell.currency.select("eur");
```

Not selecting this caused the request to be corrupt:

![CleanShot 2023-03-22 at 14 36 16@2x](https://user-images.githubusercontent.com/36263538/226921638-37dcff26-f1a6-4bb5-af08-4dc2e7f51b5b.png)
